### DISCUSSION TITLE
Replace version with digest in custom reg

### DISCUSSION THREAD

**Discussion by cloudmatt** [renovatebot/renovate#23227] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I want to pin my terraform modules to their digest as opposed to version tag since version tags are mutable. I've read the documentation and see that `terraform: {pinDigests: true}` is an option but that never actually works for me so I've attempted to go the regex route.

The tl;dr is that I want to set a custom 'comment string' that'll retain the human readable version number but update the actual module to use the digest. An example of what I'd "want" to happen looks like:

```hcl
module "aws_kms" {
-  # renovate: version=v1.0.1
-   source = "git@github.com:terraform-aws-modules/terraform-aws-kms.git?ref=v1.0.1"

+ # renovate: version=v2.0.0
+ source = "git@github.com:terraform-aws-modules/terraform-aws-kms.git?ref=MYCOOLSHADIGEST"
}
```

### What you tried so far.

Here's a [public repo](https://github.com/cloudmatt/tf-renovate-test) where I've been testing this.

I've tried the following regexManager:

```json
    {
      "description": "Search all .tf (terraform) files for a comment above source to determine how to lookup the tag to replace with the latest sha.",
      "fileMatch": ["\\.tf$"],
      "matchStrings": [
         "#\\srenovate:\\sversion=(?<currentValue>[^\n]*)\\n\\s*source.*?:(?<depName>[^\n]*)\\.git\\?ref=(?<currentVersion>[^/\"]*)\""
      ],
      "autoReplaceStringTemplate": "\\s*source.*ref={{{newDigest}}}\"",
      "datasourceTemplate": "{{#if datasource}}{{{datasource}}}{{else}}github-tags{{/if}}",
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}loose{{/if}}"
    }
```

and this opens a PR but it's only bumping the version tags and not the actual digest.

### Relevant debug logs

I've looked through the error logs and there's nothing in the ERROR or WARN that gives any insight, as far as I can tell Renovate is working like it thinks it should be.

**Comment by rarkins** [comment]
I think it would be better to get pinDigests working with terraform. It's not always possible with regex

**Reply by cloudmatt** [reply]
Is there something special needed to get that working? I've tried adding the pinDigests: true in the config but it doesn't ever pin them unless I'm doing something wrong.

**Reply by rarkins** [reply]
@secustor do you expect pinDigests to work with terraform?

**Reply by secustor** [reply]
No, it is currently not supported.

**Reply by cloudmatt** [reply]
Should [the docs](https://docs.renovatebot.com/modules/manager/terraform/#default-config) be updated to reflect that then? Because as a layperson, when I see that an option default is set as `false` the sorta 'implied' thing at that point is that `true` is an option.

**Comment by rarkins** [comment]
@cloudmatt how would you expect this to work once a pinning is done? e.g. your example is:

```
module "aws_kms" {
-  # renovate: version=v1.0.1
-   source = "git@github.com:terraform-aws-modules/terraform-aws-kms.git?ref=v1.0.1"

+ # renovate: version=v2.0.0
+ source = "git@github.com:terraform-aws-modules/terraform-aws-kms.git?ref=MYCOOLSHADIGEST"
}
```

So once the `v1.0.1` reference is "lost", how does Renovate know to keep it updated? Usually if there's a repo with digest then Renovate will keep it updated to the latest SHA in the repo.

**Reply by cloudmatt** [reply]
@rarkins - I don't think I follow what you're saying?

When renovate detects an update it knows the current value and the new value.

So in my mind the regex capture groups get:
currentVersion: 1.X.X
currentDigest: currentSha
newVersion: 2.X.X
newDigest: newSha

The all the autoreplace is doing is simply adding the digest where the regex tells it to place it, from the snippet below, which is from a repository where the digest is already set outside of Renovate, I believe renovate already has all the information it needs and from the [reproduction repo's PR](https://github.com/cloudmatt/tf-renovate-test/pull/3/files) it's able to update both 'strings' in the same manner.

```json
          {
            "currentDigest": "37b47fbde419db13bc64fb148ff485fc24c1714d",
            "currentValue": "v1.0.2",
            "currentVersion": "v1.0.2",
            "datasource": "github-releases",
            "depName": "terraform-aws-modules/terraform-aws-kms",
            "fixedVersion": "v1.0.2",
            "isSingleVersion": true,
            "packageName": "terraform-aws-modules/terraform-aws-kms",
            "registryUrl": "https://github.com",
            "replaceString": "# renovate: datasource=github-releases depName=terraform-aws-modules/terraform-aws-kms version=v1.0.2\n  source = \"git@github.com:terraform-aws-modules/terraform-aws-kms.git?ref=37b47fbde419db13bc64fb148ff485fc24c1714d\"",
            "sourceUrl": "https://github.com/terraform-aws-modules/terraform-aws-kms",
            "versioning": "semver",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "v1.5.0",
                "newValue": "v1.5.0",
                "releaseTimestamp": "2023-02-09T16:28:53.000Z",
                "newMajor": 1,
                "newMinor": 5,
                "updateType": "minor",
                "newDigest": "87be9ccb61f073b58ed56de0456214c34d793841",
                "branchName": "renovate/terraform-aws-modules-terraform-aws-kms-1.x"
              }
            ]
          },
```

**Reply by cloudmatt** [reply]
Added another test case to the reproduction repo [here](https://github.com/cloudmatt/tf-renovate-test/pull/4/files) so I think Renovate has all the information it needs to make this work? It appears it's able to track the version number as semver or whatever _and_ the digest appropriately. It would just be a matter of `/s/semver/digest/` type deal somehow

**Reply by cloudmatt** [reply]
What I do notice in the logs is that renovate is _trying_ to make a digest out of it as well but it just fails and the errors it prints aren't all that helpful unfortunately 

<details>
<summary>Debug logs</summary>

```json

DEBUG: branches info extended
{
  "branchesInformation": [
    {
      "branchName": "renovate/terraform-aws-modules-terraform-aws-kms-digest",
      "prNo": null,
      "prTitle": "chore(deps): update terraform-aws-modules/terraform-aws-kms digest to 693ec02",
      "result": "error",
      "upgrades": [
        {
          "datasource": "github-tags",
          "depName": "terraform-aws-modules/terraform-aws-kms",
          "displayPending": "",
          "fixedVersion": "v1.0.1",
          "currentVersion": "v1.0.1",
          "currentValue": "v1.0.1",
          "newValue": "v1.0.1",
          "packageFile": "main.tf",
          "updateType": "digest",
          "packageName": "terraform-aws-modules/terraform-aws-kms"
        }
      ]
    },
    {
      "branchName": "renovate/terraform-aws-modules-terraform-aws-kms-1.x",
      "prNo": 3,
      "prTitle": "chore(deps): update dependency terraform-aws-modules/terraform-aws-kms to v1.5.0",
      "result": "error",
      "upgrades": [
        {
          "datasource": "github-tags",
          "depName": "terraform-aws-modules/terraform-aws-kms",
          "displayPending": "",
          "fixedVersion": "v1.0.1",
          "currentVersion": "v1.0.1",
          "currentValue": "v1.0.1",
          "newValue": "v1.5.0",
          "newVersion": "v1.5.0",
          "packageFile": "main.tf",
          "updateType": "minor",
          "packageName": "terraform-aws-modules/terraform-aws-kms"
        },
        {
          "datasource": "github-tags",
          "depName": "terraform-aws-modules/terraform-aws-kms",
          "displayPending": "",
          "fixedVersion": "v1.0.1",
          "currentVersion": "v1.0.1",
          "currentValue": "v1.0.1",
          "newValue": "v1.5.0",
          "newVersion": "v1.5.0",
          "packageFile": "main.tf",
          "updateType": "minor",
          "packageName": "terraform-aws-modules/terraform-aws-kms"
        }
      ]
    },
    {
      "branchName": "renovate/terraform-aws-modules-terraform-aws-vpc-5.x",
      "prNo": 4,
      "prTitle": "chore(deps): update dependency terraform-aws-modules/terraform-aws-vpc to v5",
      "result": "done",
      "upgrades": [
        {
          "datasource": "github-tags",
          "depName": "terraform-aws-modules/terraform-aws-vpc",
          "displayPending": "",
          "fixedVersion": "v4.0.2",
          "currentVersion": "v4.0.2",
          "currentValue": "v4.0.2",
          "newValue": "v5.0.0",
          "newVersion": "v5.0.0",
          "packageFile": "main.tf",
          "updateType": "major",
          "packageName": "terraform-aws-modules/terraform-aws-vpc"
        }
      ]
    }
  ]
}
```

</details>

**Reply by rarkins** [reply]
I overlooked this Renovate header here:

```
+ # renovate: version=v2.0.0
+ source = "git@github.com:terraform-aws-modules/terraform-aws-kms.git?ref=MYCOOLSHADIGEST"
```

I guess you'd need this added as part of any pinDigests support. You'd want to split this into two parts:

1. Supporting the `# renovate` header to learn and be able to update `currentValue`
2. Pinning from a value to a digest

**Reply by cloudmatt** [reply]
Is there any reason why the `"autoReplaceStringTemplate":` wouldn't be able to handle that? That just does nothing for me and I'm struggling a bit on how to troubleshoot it.

**Reply by rarkins** [reply]
I'm aiming to solve this within the terraform manager, not regex

**Reply by cloudmatt** [reply]
That would be a much better solution than my chatgpt generated regex 😆

**Comment by rarkins** [comment, marked as answer]
Created #23248 and #23249
